feat: add "null" SMTP transport mode

thlehmann-ionos opened the pull request:
## Summary

If mails ought to be send by other means than rendering messages from
templates and sending them via SMTP-like protocols.


### Use-case example

Listening to specific Nextcloud events and pass parameters to a centralized (i.e. REST-based) API that sends e-mails.


### Background why not ...

* **customize mail templates:** mails should not be delivered by SMTP-like protocols and rendering messages that will never be send is unneeded overhead
* **implement template class and set `mail_template_class`:** same as before
* **listen for `BeforeMessageSent` and act upon that**: would be an option, but it would still render messages that will never be sent and the template data might also have to be enriched with custom information (i.e. in our case the sender's _user ID_, not their e-mail address)

_(Short list of internal reasoning we had before making this pull request.)_


## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- ~Screenshots before/after for front-end changes~
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- ~[Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)~

thlehmann-ionos committed:
feat(Mailer): add "null" SMTP transport mode

== Goal

Allow disabling mail delivery altogether.

== Usecase

If mails ought to be send by other means than rendering messages from
templates and sending them via SMTP-like protocols.

Example: listening to specific Nextcloud events and pass parameters to
a centralized (i.e. REST-based) API that sends e-mails.

Signed-off-by: Thomas Lehmann <t.lehmann@strato.de>

thlehmann-ionos commented:
I chose the string `null` because the `NullTransport` class in the symfony mailer package inspired me for that, but I realize, that a string `null` can be confused with a literal `null` value. I'm happs to change it to `none`, `noop` or similar.

solracsf commented on config/config.sample.php:
```diff
@@ -529,6 +529,9 @@
  * on your Unix system.
  *
+ * Use the ``null`` to send no mails (disable mail delivery). This can be useful
+ * if mails should be sent via APIs and rendering messages it not necessary.
```
```suggestion
 * if mails should be sent via APIs and rendering messages is not necessary.
```

thlehmann-ionos replied:
I will fix that. Any opinion on the term "null" vs. something different?

artonge commented on lib/private/Mail/Mailer.php:
```diff
@@ -164,6 +164,11 @@ public function createEMailTemplate(string $emailId, array $data = []): IEMailTe
 	 * @return string[] $failedRecipients
 	 */
 	public function send(IMessage $message): array {
+		if ($this->config->getSystemValueString('mail_smtpmode', 'smtp') === 'null') {
+			// The null transport indicates no mail shall be sent
+			return [];
+		}
+
```
Shouldn't we instead return the `NullTransport` here https://github.com/nextcloud-gmbh/server/blob/6969179d6be4a82cbe73d2447754a22cbb58f637/lib/private/Mail/Mailer.php#L253-L272 ?

artonge replied:
Shouldn't we keep the event dispatching?

thlehmann-ionos replied:
For consistency yes. Changed.

thlehmann-ionos committed:
feat(Mailer): implement caching

Currently $this->instance is never set, so the code is no-op. This
brings back caching of the instance.

Caching broke with

   be7db1573dc8c6e7309ec9db124a7a74b8b41199

   Swift to \Swift_Mailer as abstraction

Signed-off-by: Thomas Lehmann <t.lehmann@strato.de>

thlehmann-ionos committed:
chore(Mailer): remove no-op local variable initilization

The IDE hinted the value is immediately overwritten.

Signed-off-by: Thomas Lehmann <t.lehmann@strato.de>

come-nc reviewed (commented):
Current version looks good, it adds support for NullTransport.

How will the SMTP settings UI behave when using null transport though?

artonge commented on lib/private/Mail/Mailer.php:
```diff
@@ -251,13 +250,10 @@ public function validateMailAddress(string $email): bool {
 	}
 
 	protected function getInstance(): MailerInterface {
-		if (!is_null($this->instance)) {
-			return $this->instance;
-		}
```
Why drop the caching?

thlehmann-ionos replied:
It's never written currently.

Apparently it broke with be7db1573dc8c6e7309ec9db124a7a74b8b41199.

thlehmann-ionos replied:
Should I create a follow-up issue to add caching back and leave the clean-up commit  or drop the cleanup commit?

artonge replied:
Ah, good catch. I feel like it was an oversight to not assign `$transport` to `$this->instance`. However, I do not know how useful the caching is.

As you worked on it, can the `getInstance` method be called multiple time during the same request? And if so, is it costly?

In doubt, let's re-introduce the caching :).

artonge commented on tests/lib/Mail/MailerTest.php:
```diff
@@ -114,6 +115,28 @@ public function testGetSendmailInstanceSendMailQmail($sendmailMode, $binaryParam
+			->with('mail_smtpmode', 'smtp')
+			->willReturn('null');
+
+		// $message = $this->createMock(Message::class);
```
Can be remove

artonge reviewed (approved):
Looks good :)

thlehmann-ionos committed:
feat(Mailer): hide admin settings for "null" transport

When the mail transport is configured as null transport, the
configuration UI would not work.

== Background

The null transport is meant for situations where operators
implement mail delivery via custom mechanisms like REST APIs.

Signed-off-by: Thomas Lehmann <t.lehmann@strato.de>

artonge commented:
@thlehmann-ionos, php linter is not happy, can you fix it? You probably just need to run `composer run cs:check`

Ch3ck3rM0n commented:
Hi all, I have installed the current "31.0.0 beta 4". I read some comments and documentations and disabled the smtpmode in the config.php, but I am still have warnings due to email configuration in the overview.

```
  'mail_smtpmode' => 'null',
  'mail_sendmailmode' => 'null',
```

Did I miss something?

![image](https://github.com/user-attachments/assets/5e84fe67-bacf-4255-b911-f2f4c4b21fd5)

come-nc commented:
> Hi all, I have installed the current "31.0.0 beta 4". I read some comments and documentations and disabled the smtpmode in the config.php, but I am still have warnings due to email configuration in the overview.
> 
> ```
>   'mail_smtpmode' => 'null',
>   'mail_sendmailmode' => 'null',
> ```
> 
> Did I miss something?
> 
> ![image](https://private-user-images.githubusercontent.com/102294555/404919534-5e84fe67-bacf-4255-b911-f2f4c4b21fd5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczODg1OTMsIm5iZiI6MTczNzM4ODI5MywicGF0aCI6Ii8xMDIyOTQ1NTUvNDA0OTE5NTM0LTVlODRmZTY3LWJhY2YtNDI1NS1iOTExLWYyZjRjNGIyMWZkNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyMFQxNTUxMzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05N2ZkZTk5YWE1YzcyZWVjYmFiMDM1Njc3MWNlZjIyZmNlOTAxYzU1M2E4OGI1Y2FhYzczNWQyOWFhZGQwNmZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.aYGlnpecSSX_EJ38cgMmBLcVg9waZEni9C_3RTMLrDw)

No you did not, it was missed when adding `null` that the setup check should be adapted as well.
I will fix it for next version.
